v2/report_publisher.go

```diff
@@ -1,27 +1,106 @@
 package bugsnag
 
-import "fmt"
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+)
 
 type reportPublisher interface {
 	publishReport(*payload) error
+	setMainProgramContext(context.Context)
 }
 
-type defaultReportPublisher struct{}
+var mainProgramDone = false
 
-func (*defaultReportPublisher) publishReport(p *payload) error {
+func (defPub *defaultReportPublisher) delivery() {
+	signalsCh := make(chan os.Signal, 1)
+	signal.Notify(signalsCh, syscall.SIGINT, syscall.SIGTERM)
+	closeDelivery := false
+
+waitForEnd:
+	for {
+		select {
+		case <-signalsCh:
+			mainProgramDone = true
+			break waitForEnd
+		case <-defPub.mainProgramCtx.Done():
+			mainProgramDone = true
+			break waitForEnd
+		case p, ok := <-defPub.eventsChan:
+			if ok {
+				if err := p.deliver(); err != nil {
+					// Ensure that any errors are logged if they occur in a goroutine.
+					p.logf("bugsnag/defaultReportPublisher.publishReport: %v", err)
+				}
+			} else {
+				p.logf("Event channel closed")
+				return
+			}
+		}
+	}
+
+	// Send remaining elements from the queue
+	for !closeDelivery {
+		select {
+		case p, ok := <-defPub.eventsChan:
+			if ok {
+				if err := p.deliver(); err != nil {
+					// Ensure that any errors are logged if they occur in a goroutine.
+					p.logf("bugsnag/defaultReportPublisher.publishReport: %v", err)
+				}
+			} else {
+				p.logf("Event channel closed")
+				return
+			}
+		default:
+			if mainProgramDone && len(defPub.eventsChan) == 0 {
+				fmt.Println("Main program done and event channel empty")
+				closeDelivery = true
+			}
+		}
+	}
```

Comment:
```suggestion
    close(defPub.eventsChan)
    for p := range defPub.eventsChan {
        if err := p.deliver(); err != nil {
	        // Ensure that any errors are logged if they occur in a goroutine.
	        p.logf("bugsnag/defaultReportPublisher.publishReport: %v", err)
        }
    }
```

Reply:
I think it won't work - the producer is supposed to close the channel, after it's closed the "for range" won't read anything.
https://go.dev/tour/concurrency/4

> The loop for i := range c receives values from the channel repeatedly until it is closed.